thetis/callback.py

```diff
@@ -5,35 +5,152 @@
 """
 from __future__ import absolute_import
 from .utility import *
+from abc import ABCMeta, abstractproperty, abstractmethod
+import h5py
 
 
-class ValidationCallback(object):
+class CallbackManager(object):
     """
-    Base class of all objects used to check validity of simulation
-    at each export.
+    Stores callbacks in different categories and provides methods for
+    evaluating them.
     """
-    def __init__(self):
-        self._initialized = False
+    def __init__(self, modes=('export', 'timestep')):
+        self.callbacks = {}
+        for mode in modes:
+            self.callbacks[mode] = OrderedDict()
 
-    def initialize(self, comm=COMM_WORLD):
-        """Initializes the callback function.
+    def add(self, callback, mode):
+        key = callback.name
+        self.callbacks[mode][key] = callback
 
-        :kwarg comm: The communicator for the callback."""
-        self._initialized = True
+    def evaluate(self, solver_obj, mode):
+        for key in self.callbacks[mode]:
+            self.callbacks[mode][key].dostuff(solver_obj)
+
+    def __getitem__(self, key):
+        return self.callbacks[key]
+
+
+class DiagnosticHDF5(object):
+    """Creates hdf5 files for diagnostic time series arrays."""
+    def __init__(self, filename, varnames, comm=COMM_WORLD):
+        self.comm = comm
+        self.filename = filename
+        self.varnames = varnames
+        self.nvars = len(varnames)
+        if comm.rank == 0:
+            # create empty file with correct datasets
+            hdf5file = h5py.File(filename, 'w')
+            dataarrays = {}
```

Comment:
It was used in my first implementation, not really needed anymore. I'll remove it.